File: packages/npm/jns42-generator/src/schema-transforms/resolve-all-of.ts
import { SchemaItem, SchemaTransform, intersectionMergeTypes } from "../models/index.js";
import {
  booleanMergeAnd,
  booleanMergeOr,
  intersectionMerge,
  mergeKeysArray,
  mergeKeysRecord,
  numericMergeMaximum,
  numericMergeMinimum,
  numericMergeMultipleOf,
  unionMerge,
} from "../utils/index.js";

/**
 * This transformer merges all sub schemas in allOf.
 *
 * ```yaml
 * - allOf
 *   - required: ["a", "b"]
 *     objectProperties:
 *       a: 100
 *       b: 200
 *   - objectProperties:
 *       b: 300
 *       c: 400
 * ```
 *
 * will become
 *
 * ```yaml
 * - required: ["a", "b"]
 *   objectProperties:
 *     a: 100
 *     b: 1
 *     c: 400
 * - allOf
 *   - 200
 *   - 300
 * ```
 */
export const resolveAllOf: SchemaTransform = (arena, key) => {
  const item = arena.getItem(key);

  // we need at least two to merge
  if (item.allOf == null || item.allOf.length < 2) {
    return;
  }

  if (item.types != null) {
    return;
  }

  if (item.reference != null) {
    return;
  }

  if (item.if != null) {
    return;
  }

  if (item.then != null) {
    return;
  }

  if (item.else != null) {
    return;
  }

  if (item.not != null) {
    return;
  }

  if (item.oneOf != null && item.oneOf.length > 0) {
    return;
  }

  if (item.anyOf != null && item.anyOf.length > 0) {
    return;
  }

  let itemNew: SchemaItem = {
    ...item,
    allOf: undefined,
  };

  for (const subKey of item.allOf) {
    let subItem = arena.getItem(subKey);

    if (subItem.types != null && subItem.types.length > 1) {
      return;
    }

    if (subItem.reference != null) {
      return;
    }

    if (subItem.if != null) {
      return;
    }

    if (subItem.then != null) {
      return;
    }

    if (subItem.else != null) {
      return;
    }

    if (subItem.not != null) {
      return;
    }

    if (subItem.oneOf != null && subItem.oneOf.length > 0) {
      return;
    }

    if (subItem.anyOf != null && subItem.anyOf.length > 0) {
      return;
    }

    if (subItem.allOf != null && subItem.allOf.length > 0) {
      return;
    }

    itemNew = {
      ...itemNew,

      // every item needs to be exact
      exact: booleanMergeAnd(itemNew.exact, subItem.exact),

      types: intersectionMergeTypes(itemNew.types, subItem.types),
      options: intersectionMerge(itemNew.options, subItem.options),
      required: unionMerge(itemNew.required, subItem.required),
      propertyNames: mergeKey(itemNew.propertyNames, subItem.propertyNames),
      contains: mergeKey(itemNew.contains, subItem.contains),
      tupleItems: mergeKeysArray(itemNew.tupleItems, subItem.tupleItems, mergeKey),
      arrayItems: mergeKey(itemNew.arrayItems, subItem.arrayItems),
      objectProperties: mergeKeysRecord(
        itemNew.objectProperties,
        subItem.objectProperties,
        mergeKey,
      ),
      mapProperties: mergeKey(itemNew.mapProperties, subItem.mapProperties),

      minimumInclusive: numericMergeMinimum(itemNew.minimumInclusive, subItem.minimumInclusive),
      minimumExclusive: numericMergeMinimum(itemNew.minimumExclusive, subItem.minimumExclusive),
      maximumInclusive: numericMergeMaximum(itemNew.maximumInclusive, subItem.maximumInclusive),
      maximumExclusive: numericMergeMaximum(itemNew.maximumExclusive, subItem.maximumExclusive),
      multipleOf: numericMergeMultipleOf(itemNew.multipleOf, subItem.multipleOf),
      minimumLength: numericMergeMinimum(itemNew.minimumLength, subItem.minimumLength),
      maximumLength: numericMergeMaximum(itemNew.maximumLength, subItem.maximumLength),
      valuePattern: unionMerge(itemNew.valuePattern, subItem.valuePattern),
      valueFormat: unionMerge(itemNew.valueFormat, subItem.valueFormat),
      minimumItems: numericMergeMinimum(itemNew.minimumItems, subItem.minimumItems),
      maximumItems: numericMergeMaximum(itemNew.maximumItems, subItem.maximumItems),
      uniqueItems: booleanMergeOr(itemNew.uniqueItems, subItem.uniqueItems),
      minimumProperties: numericMergeMinimum(itemNew.minimumProperties, subItem.minimumProperties),
      maximumProperties: numericMergeMaximum(itemNew.maximumProperties, subItem.maximumProperties),
    };
  }

  arena.setItem(key, itemNew);

  function mergeKey(key: number | undefined, otherKey: number | undefined): number | undefined {
    if (key === otherKey) {
      return key;
    }

    if (key == null) {
      return otherKey;
    }
    if (otherKey == null) {
      return key;
    }

    const itemNew = {
      allOf: [key, otherKey],
    };
    const keyNew = arena.addItem(itemNew);
    return keyNew;
  }
};

File: packages/cargo/jns42-generator/src/schema_transforms/resolve_all_of.rs
use crate::models::{arena::Arena, schema::SchemaNode};
use std::cell::RefCell;

/**
 * This transformer merges all sub schemas in allOf.
 *
 * ```yaml
 * - allOf
 *   - required: ["a", "b"]
 *     objectProperties:
 *       a: 100
 *       b: 200
 *   - objectProperties:
 *       b: 300
 *       c: 400
 * ```
 *
 * will become
 *
 * ```yaml
 * - required: ["a", "b"]
 *   objectProperties:
 *     a: 100
 *     b: 1
 *     c: 400
 * - allOf
 *   - 200
 *   - 300
 * ```
 */
pub fn transform(arena: &mut Arena<SchemaNode>, key: usize) {
  let item = arena.get_item(key);

  let Some(sub_keys) = item.all_of.clone() else {
    return;
  };

  if sub_keys.len() < 2 {
    return;
  };

  // things we cannot merge
  if item
    .types
    .as_ref()
    .map(|value| value.len())
    .unwrap_or_default()
    > 1
    || item.reference.is_some()
    || item.any_of.is_some()
    || item.one_of.is_some()
    || item.r#if.is_some()
    || item.then.is_some()
    || item.r#else.is_some()
    || item.not.is_some()
  {
    return;
  }

  let mut item_new = SchemaNode {
    all_of: None,
    ..item.clone()
  };

  // this is so that the following closures don't have to be FnMut
  let arena = RefCell::new(arena);

  let merge_key = |key: &usize, other_key: &usize| {
    if key == other_key {
      return *key;
    }

    let item_new = SchemaNode {
      all_of: Some([*key, *other_key].into()),
      ..Default::default()
    };

    arena.borrow_mut().add_item(item_new)
  };

  for sub_key in sub_keys {
    let arena = arena.borrow_mut();
    let sub_item = arena.get_item(sub_key);

    // things we cannot merge
    if sub_item
      .types
      .as_ref()
      .map(|value| value.len())
      .unwrap_or_default()
      > 1
      || sub_item.reference.is_some()
      || sub_item.all_of.is_some()
      || sub_item.any_of.is_some()
      || sub_item.one_of.is_some()
      || sub_item.r#if.is_some()
      || sub_item.then.is_some()
      || sub_item.r#else.is_some()
      || sub_item.not.is_some()
    {
      return;
    }

    item_new = item_new.intersection(sub_item, &merge_key);
  }

  let arena = arena.into_inner();

  arena.replace_item(key, item_new);
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::models::{
    arena::Arena,
    schema::{SchemaNode, SchemaType},
  };

  #[test]
  fn test_utility() {
    let mut arena = Arena::from_iter([
      SchemaNode {
        types: Some([SchemaType::Never].into()),
        ..Default::default()
      }, // 0
      SchemaNode {
        types: Some([SchemaType::Any].into()),
        ..Default::default()
      }, // 1
      SchemaNode {
        types: Some([SchemaType::Number].into()),
        ..Default::default()
      }, // 2
      SchemaNode {
        all_of: Some([2, 0].into()),
        ..Default::default()
      }, // 3
      SchemaNode {
        all_of: Some([2, 1].into()),
        ..Default::default()
      }, // 4
    ]);

    while arena.apply_transform(transform) > 0 {
      //
    }

    let actual: Vec<_> = arena.iter().cloned().collect();
    let expected: Vec<_> = [
      SchemaNode {
        types: Some([SchemaType::Never].into()),
        ..Default::default()
      }, // 0
      SchemaNode {
        types: Some([SchemaType::Any].into()),
        ..Default::default()
      }, // 1
      SchemaNode {
        types: Some([SchemaType::Number].into()),
        ..Default::default()
      }, // 2
      SchemaNode {
        types: Some([SchemaType::Never].into()),
        ..Default::default()
      }, // 3
      SchemaNode {
        types: Some([SchemaType::Number].into()),
        ..Default::default()
      }, // 4
    ]
    .into();

    assert_eq!(actual, expected)
  }

  #[test]
  fn test_primitive() {
    let mut arena = Arena::from_iter([
      SchemaNode {
        types: Some([SchemaType::Number].into()),
        ..Default::default()
      }, // 0
      SchemaNode {
        types: Some([SchemaType::String].into()),
        ..Default::default()
      }, // 1
      SchemaNode {
        types: Some([SchemaType::String].into()),
        ..Default::default()
      }, // 2
      SchemaNode {
        all_of: Some([0, 1].into()),
        ..Default::default()
      }, // 3
      SchemaNode {
        all_of: Some([1, 2].into()),
        ..Default::default()
      }, // 4
    ]);

    while arena.apply_transform(transform) > 0 {
      //
    }

    let actual: Vec<_> = arena.iter().cloned().collect();
    let expected: Vec<_> = [
      SchemaNode {
        types: Some([SchemaType::Number].into()),
        ..Default::default()
      }, // 0
      SchemaNode {
        types: Some([SchemaType::String].into()),
        ..Default::default()
      }, // 1
      SchemaNode {
        types: Some([SchemaType::String].into()),
        ..Default::default()
      }, // 2
      SchemaNode {
        types: Some([SchemaType::Never].into()),
        ..Default::default()
      }, // 3
      SchemaNode {
        types: Some([SchemaType::String].into()),
        ..Default::default()
      }, // 4
    ]
    .into();

    assert_eq!(actual, expected)
  }

  #[test]
  fn test_tuple() {
    let mut arena = Arena::from_iter([
      SchemaNode {
        types: Some([SchemaType::Number].into()),
        ..Default::default()
      }, // 0
      SchemaNode {
        types: Some([SchemaType::String].into()),
        ..Default::default()
      }, // 1
      SchemaNode {
        types: Some([SchemaType::String].into()),
        ..Default::default()
      }, // 2
      SchemaNode {
        types: Some([SchemaType::String].into()),
        ..Default::default()
      }, // 3
      SchemaNode {
        types: Some([SchemaType::Array].into()),
        tuple_items: Some([0, 1].into()),
        ..Default::default()
      }, // 4
      SchemaNode {
        types: Some([SchemaType::Array].into()),
        tuple_items: Some([2, 3].into()),
        ..Default::default()
      }, // 5
      SchemaNode {
        all_of: Some([4, 5].into()),
        ..Default::default()
      }, // 6
    ]);

    while arena.apply_transform(transform) > 0 {
      //
    }

    let actual: Vec<_> = arena.iter().cloned().collect();
    let expected: Vec<_> = [
      SchemaNode {
        types: Some([SchemaType::Number].into()),
        ..Default::default()
      }, // 0
      SchemaNode {
        types: Some([SchemaType::String].into()),
        ..Default::default()
      }, // 1
      SchemaNode {
        types: Some([SchemaType::String].into()),
        ..Default::default()
      }, // 2
      SchemaNode {
        types: Some([SchemaType::String].into()),
        ..Default::default()
      }, // 3
      SchemaNode {
        types: Some([SchemaType::Array].into()),
        tuple_items: Some([0, 1].into()),
        ..Default::default()
      }, // 4
      SchemaNode {
        types: Some([SchemaType::Array].into()),
        tuple_items: Some([2, 3].into()),
        ..Default::default()
      }, // 5
      SchemaNode {
        types: Some([SchemaType::Array].into()),
        tuple_items: Some([7, 8].into()),
        ..Default::default()
      }, // 6
      SchemaNode {
        types: Some([SchemaType::Never].into()),
        ..Default::default()
      }, // 7
      SchemaNode {
        types: Some([SchemaType::String].into()),
        ..Default::default()
      }, // 8
    ]
    .into();

    assert_eq!(actual, expected)
  }

  #[test]
  fn test_array() {
    let mut arena = Arena::from_iter([
      SchemaNode {
        types: Some([SchemaType::Number].into()),
        ..Default::default()
      }, // 0
      SchemaNode {
        types: Some([SchemaType::String].into()),
        ..Default::default()
      }, // 1
      SchemaNode {
        types: Some([SchemaType::Array].into()),
        array_items: Some(0),
        ..Default::default()
      }, // 2
      SchemaNode {
        types: Some([SchemaType::Array].into()),
        array_items: Some(1),
        ..Default::default()
      }, // 3
      SchemaNode {
        all_of: Some([2, 3].into()),
        ..Default::default()
      }, // 4
    ]);

    while arena.apply_transform(transform) > 0 {
      //
    }

    let actual: Vec<_> = arena.iter().cloned().collect();
    let expected: Vec<_> = [
      SchemaNode {
        types: Some([SchemaType::Number].into()),
        ..Default::default()
      }, // 0
      SchemaNode {
        types: Some([SchemaType::String].into()),
        ..Default::default()
      }, // 1
      SchemaNode {
        types: Some([SchemaType::Array].into()),
        array_items: Some(0),
        ..Default::default()
      }, // 2
      SchemaNode {
        types: Some([SchemaType::Array].into()),
        array_items: Some(1),
        ..Default::default()
      }, // 3
      SchemaNode {
        types: Some([SchemaType::Array].into()),
        array_items: Some(5),
        ..Default::default()
      }, // 4
      SchemaNode {
        types: Some([SchemaType::Never].into()),
        ..Default::default()
      }, // 5
    ]
    .into();

    assert_eq!(actual, expected)
  }
}
